Initial Entity and Resource proposal. (#264)
open-telemetry/oteps: text/entities/0264-resource-and-entities.md

Going forward, we have set of options:

- Duplicate attributes in `Entity` section of Resource.
- Reference attributes of Resource in entity.
- Only identify Entity id and keep attribute<->entity association out of band.
- Extend Attribute on Resource so that we can track the entity type per Key-Value (across any attribute in OTLP).

The third option prevents generic code from interacting with Resource and Entity without understanding the model of each.  The first keeps all usage of entity simple at the expense of duplicating information and the middle is awkward to interact with from an OTLP usage perspective. The fourth is violates our stability policy for OTLP.

## How to deal with Resource/Entities whose lifecycle does not match the SDK?

This proposal motivates a Resource Coordinator in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.

## How to deal with Prometheus Compatibility for non-SDK telemetry?

> Today, Prometheus compatibility relies on two key attributes in Resource: service.name and service.instance.id. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.

Reviewer: Based on reading through this, here is what I think would make the most sense:

* `target_info` metric generation is updated to exclude any keys which are contained in `descriptive_attributes_keys` of an entity.
* For each entity which has non-empty `descriptive_attributes_keys`, generate an info metric: `<entity>_entity_info` (naming TBD), which has all identifying and descriptive keys.  This should play nicely with the planned improvements to info-typed metrics in https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals

> Persist info metrics with labels categorized as either identifying or non-identifying (i.e. data labels).

Reviewer: I think you should include the first bullet in this proposal (excluding descriptive attributes from target_info).  It should be non-controversial.  The second bullet can be proposed separately, since it is additive.

Author: Added this comment, thanks!

## Design

### Approach - Resource Improvements

Let's focus on outlining Entity detectors and Resource composition.  This has a higher priority for fixing within OpenTelemetry, and needs to be unblocked sooner. Then infer our way back to data model and Collector use cases.

We define the following SDK components:

- **Resource Detectors (legacy)**:  We preserve existing resource detectors.  They have the same behavior and interfaces as today.
- **Entity Detectors (new)**: Detecting an entity that is relevant to the current instance of the SDK.  For example, this would detect a service entity for the current SDK, or its process. Every entity must have some relation to the current SDK.
- **Resource Coordinator (new)**: A component responsible for taking Resource and Entity detectors and doing the following:
  - Constructing a Resource for the SDK from detectors.
  - Dealing with conflicts between detectors.
  - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.
  - *(new) Managing Entity changes during SDK lifetime, specifically dealing with entities that have lifetimes shorter than the SDK*

> #### Resource Container

Reviewer 2: Is this meant to say resource coordinator or is resource container different?

Author: Yeah - too aggressive in the auto-spellcheck

> The SDK Resource coordinator is responsible for running all configured Resource and Entity Detectors.  There will be some (user-controlled, otel default) priority order to these.

Reviewer 3: Is there an example of what the default priority would be?

Author: I'd like to let distributions decide, but let's say you're running on GCP - I'd imagine an order like:

```
entity_detectors:
  - env
  - gcp
  - K8S
  - process
  - sdk
```

I'd like to sort this out when writing the actual specification as I'd like to implement this across several languages and runtimes to get a feel for how flexible the language needs to be to support SDKs.

- Allow semantic convention resource modeling to progress ([spec#605](https://github.com/open-telemetry/opentelemetry-specification/issues/605), [spec#559](https://github.com/open-telemetry/opentelemetry-specification/issues/559), etc).

## Design

### Approach - Resource Improvements

Let's focus on outlining Entity detectors and Resource composition.  This has a higher priority for fixing within OpenTelemetry, and needs to be unblocked sooner. Then infer our way back to data model and Collector use cases.

We define the following SDK components:

- **Resource Detectors (legacy)**:  We preserve existing resource detectors.  They have the same behavior and interfaces as today.
- **Entity Detectors (new)**: Detecting an entity that is relevant to the current instance of the SDK.  For example, this would detect a service entity for the current SDK, or its process. Every entity must have some relation to the current SDK.
- **Resource Coordinator (new)**: A component responsible for taking Resource and Entity detectors and doing the following:
  - Constructing a Resource for the SDK from detectors.
  - Dealing with conflicts between detectors.

>   - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.

Author: The TL;DR; here is that we agreed "Entity" signal would be an Event/Log.  The idea here is that you could export discovered Entities or entity change events via logs in the SDK here.

I can remove this as it's not that important, but something we can add later if we decide entities should be exported from SDKs + their detection.

>   - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.

Reviewer 4: I assume using logs [like this](https://docs.google.com/document/d/1Tg18sIck3Nakxtd3TFFcIjrmRO_0GLMdHXylVqBQmJA/edit#heading=h.v4ilwdkncxe)?

>   - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.

Author: Yes.  I suspect we'll need to look at what cloud events is doing here as well, but I don't think we need this functionality in an initial implementation.  This sets up our architecture to support that feature (which I suspect we will need, but can defer).

>   - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.

Reviewer 5: Second that this will be needed, as appending the full state of entities to all outgoing telemetry can be quite expensive, as can managing partial updates on the back-end. Creating a separate events stream with full entity state can relieve these pressures.

>   - Providing SDK-internal access to detected Resources for reporting via Log signal on configured LogProviders.

Author: @HunterSherms Absolutely agree.  The important point here is that this MAY happen directly from SDKs and not just within an OTEL collector, e.g.

  - *(new) Managing Entity changes during SDK lifetime, specifically dealing with entities that have lifetimes shorter than the SDK*

#### Resource Coordinator

> The SDK Resource Coordinator is responsible for running all configured Resource and Entity Detectors.  There will be some (user-controlled, otel default) priority order to these.

Reviewer 6: It sounds we need to document the set of known entity types, their associated attributes, and their relative priority. This seems like something for the `semantic-conventions` repo eventually. Is that what you had in mind?

Reviewer 4: >It sounds we need to document the set of known entity types, their associated attributes ... This seems like something for the semantic-conventions repo eventually. Is that what you had in mind?

Yes, that's the plan.

> ... and their relative priority. 

That's a good question. I was thinking it will be an SDK configuration chosen by the end user, but perhaps it has a place in semconv?

Author: We MUST expose it to end users, but I think I agree, for OTEL itself we should have a default for these things.

I'd like to expand this as an open question for how to solve, but semconv is a likely candidate for that.

Author: If you want to understand what semconv already have - I just migrated `type` into the YAML here: https://github.com/open-telemetry/semantic-conventions/pull/1400

It's a good survey of what our community was building prior to this OTEP.

- The resource coordinator will detect conflicts in Entity of the same type being discovered and choose one to use.
- When using Entity Detectors and Resource detectors together, the following merge rules will be used:
  - Entity merging will occur first resulting in an "Entity Merged" Resource.
    - Entities of different types will be merged into the resulting Resource.

>     - Entities of the same type will have one rejected and one accepted, based on priority.

Author: Yes - I think we can merge if the identities are the same, and add descriptive attributes.  I can update this appropriately, or if you have a proposal I'll add it directly here.

>     - Entities of the same type will have one rejected and one accepted, based on priority.

Author: Let me flesh out this section more given comments, but a quick answer to identity:

Since Entity can have descriptive attributes, when two entities have the same type, we can merge descriptive attributes.

Regarding schema_url -> Nominally when merging two of the "same" entities, we'd need to execute schema_url translations to do so, e.g. see: https://github.com/open-telemetry/oteps/pull/161

So you're right about identity, but also I think we have two "identities" to think about here:

- When are two entities describing the same thing and can be trivially merged
- When are two entities describing the same thing and can be merged by interacting with schema_url

I'm not sure how to talk about this "simply" yet, so I'll work on expanding the merge algorithm to show case this, and be more explicit about what fields are used for "identity" in each case.

>     - Entities of the same type will have one rejected and one accepted, based on priority.

Reviewer 4: >Regarding schema_url -> Nominally when merging two of the "same" entities, we'd need to execute schema_url translations to do so, e.g. see

+1.

  - Resource detectors otherwise follow existing merge semantics.
    - The Specification merge rules will be updated to account for violations prevalent in ALL implementation of resource detection.

>     - Specifically: This means the rules around merging Resource across schema-url will be dropped.  Instead only conflicting attributes will be dropped.

Reviewer 4: Please add that link to the OTEP text.

>     - Specifically: This means the rules around merging Resource across schema-url will be dropped.  Instead only conflicting attributes will be dropped.

Author: done

>     - SchemaURL on Resource will need to be deprecated with entity-specific schema-url replacing it.  Additionally, as no Resource semantic conventions have ever stabilized, SchemaURL usage on Resource cannot be in stable components of OpenTelemetry.  Given prevalent violation of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal.

Author: Yes.  Given most implementations of SDK violate this, we can't really trust SchemaURL as-is or they are explicitly dropping it today.

>     - SchemaURL on Resource will need to be deprecated with entity-specific schema-url replacing it.  Additionally, as no Resource semantic conventions have ever stabilized, SchemaURL usage on Resource cannot be in stable components of OpenTelemetry.  Given prevalent violation of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal.

Author: Yes

>     - SchemaURL on Resource will need to be deprecated with entity-specific schema-url replacing it.  Additionally, as no Resource semantic conventions have ever stabilized, SchemaURL usage on Resource cannot be in stable components of OpenTelemetry.  Given prevalent violation of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal.

Reviewer 4: SGTM. Let's make it clear in the OTEP text.

>     - SchemaURL on Resource will need to be deprecated with entity-specific schema-url replacing it.  Additionally, as no Resource semantic conventions have ever stabilized, SchemaURL usage on Resource cannot be in stable components of OpenTelemetry.  Given prevalent violation of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal.

Author: added

>   - An OOTB "Env Variable Entity Detector" will be specified and provided vs. requiring SDK wide ENV variables for resource detection.

Reviewer 6: Wasn't clear to me that this detector was elaborated on below. Maybe provide a link?

Author: done

- *Additionally, Resource Coordinator would be responsible for understanding Entity lifecycle events, for Entities whose lifetimes do not match or exceed the SDK's own lifetime (e.g. browser session).*

#### Entity Detector

The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK.  For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod.   SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

An Entity Detector would have an API similar to:

```rust
trait EntityDetector

>   pub fn detect_entities(...) -> Result<Vec<Entity>, EntityDetectionError>>

Reviewer 6: Are the descriptive and identifying attributes encapsulated in `Entity` in this example?

Author: yes

```

Where `Result` is the equivalent of error channel in the language of choice (e.g. in Go this would be `entities, err := e.detectEntities()`).

#### Entity Merging and Resource

The most important aspect of this design is how Entities will be merged to construct a Resource. We provide a simple algorithm for this behavior:

- Construct a set of detected entities, E

> - All entity detectors are sorted by priority

Reviewer 6: ```suggestion
- All entity detectors are sorted by priority (highest first)
```

- For each entity detector
  - For each entity detected

>     - If the entity exists in E, ignore it

Reviewer 4: To clarify: "If the entity exists in E" will use entity's (`Type`, `Id`) field pair for presence detection in set `E`, right? The mutable `Attributes` field of the entities are ignored in set `E` construction.

Author: Yes - we use identity of the entity to see if it exists.

    - Otherwise, add the entity to E
- Construct a Resource from the set E.

> Any implementation that achieves the same result as this algorithm is acceptable.

Reviewer 6: What do you mean by this sentence? Do you imagine efficiency problems with this that would suggest alternative implementations? What is accomplished by leaving the door open to alternatives?

Author: The algorithm as defined is very imperative/top down.  I can imagine bottom up implementations or others.  I don't think we gain anything by forcing an algorithm, and I do expect folks could improve upon this.

> #### Environment Variable Detector

Author: The goal of this *is* to interact with ENV variables directly, i.e. that's not configurable.  See the second part of this comment: https://github.com/open-telemetry/opentelemetry-specification/issues/3966#issuecomment-2037126970

I think the interaction with config is the *priority* of when a resource attribute is used, but the actual names and things need to be specified and consistent to enable things like otel-operator (or platform providers) to consistently offer entity information without needing to have custom code in every language.

> #### Environment Variable Detector

Reviewer 7: nice! this is the solution to the problem we had in Azure Functions with file config!

> An Entity detector will be specified to allow Platform to inject entity identity information into workloads running on that platform.   For Example, the OpenTelemetry Operator could inject information about Kubernetes Deployment + Container into the environment, which SDKs can elect to interact with (through configuration of the Environment Variable Entity Detector).

Reviewer 4: Do we have a definition of "Platform" anywhere? It is not in spec's glossary.

Author: It's used in one of the issues: https://github.com/open-telemetry/opentelemetry-specification/issues/3966

We'll have to add this to the spec glossary I assume, but we don't have the use cases for Platform called out today.  This is addressing that for Enttiy/Resource.

Reviewer 4: To make it clear it would be useful to add a quick definition in this OTEP.

Author: done.

While details of ENV variables will be subject to change, it would look something like the following:

```bash
set OTEL_DETECTED_ENTITIES=k8s.deployment[k8s.deployment.name=my-program],k8s.pod[k8s.pod.name=my-program-2314,k8s.namespace=default]
<run my program>
```

The minimum requirements of this entity detector are:

- ENV variable can specify multiple entities (resource attribute bundles)

> - ENV variable can be easily appended or leverages by multiple participating systems, if needed.

Reviewer 6: And influence priority if done right.. I.e. the difference between
```
export OTEL_DETECTED_ENTITIES=${OTEL_DETECTED_ENTITIES},${MY_PLATFORM_ENTITIES}
```
and
```
export OTEL_DETECTED_ENTITIES=${MY_PLATFORM_ENTITIES},${OTEL_DETECTED_ENTITIES},
```

Author: Love it, yes - This should have been a requirement all along.

> - Entities discovered via ENV variable can participate in Resource Manager generically, i.e. resolving conflicting definitions.

Reviewer 4: Is Resource Manager different from Resource Coordinator?

Author: No, that's a typo.

The actual design for this ENV variable interaction would follow the approval of this OTEP.

### Interactions with OpenTelemetry Collector

The OpenTelemetry collector can be updated to optionally  interact with Entity on Resource. A new entity-focused resource detection process can be created which allows add/override behavior at the entity level, rather than individual attribute level.

For example, the existing resource detector looks like this:

```yaml
processors:
  resourcedetection/docker:
    detectors: [env, docker]
    timeout: 2s
    override: false
```

The future entity-based detector would look almost exactly the same, but interact with the entity model of resource:

```yaml
processor:

>   entityresourcedetection:

Reviewer 4: I think instead of introducing a new `entityresourcedetection` processor we should aim to modify `resourcedetection` processor in backward compatible manner (preserving it's old behavior of Resource detection) and add entity detection capabilities. See how it's done in this prototype: https://github.com/tigrannajaryan/opentelemetry-collector-contrib/pull/1/files

     # Order determines override behavior
     detectors: [env, docker]
     # False means only append if entity doesn't already exist.
     override: false 
```

The list of detectors is given in priority order (first wins, in event of a tie, outside of override configuration). The processor may need to be updated to allow the override flag to apply to each individual detector.

## Datamodel Changes

Given our desired design and algorithms for detecting, merging and manipulating Entities, we need the ability to denote how entity and resource relate. These changes must not break existing usage of Resource, therefore:

> - The Entity model must be *layered on top of* the Resource model.  A system does not need to ineract with entities for correct behavior.

Reviewer 4: ```suggestion
- The Entity model must be *layered on top of* the Resource model.  A system does not need to interact with entities for correct behavior.
```

| dropped_attributes_count |  integer | dropped_attributes_count is the number of dropped attributes. If the value is 0, then no attributes were dropped. | Unchanged |
| attributes | repeated KeyValue | Set of attributes that describe the resource.<br/><br/>Attribute keys MUST be unique (it is not allowed to have more than one attribute with the same key).| Unchanged |
| entities | repeated ResourceEntityRef | Set of entities that participate in this Resource. | Added |

The DataModel would ensure that attributes in Resource are produced from both the identifying and descriptive attributes of Entity.  This does not mean the protocol needs to transmit duplicate data, that design is TBD.

### ResourceEntityRef

The entityref data model, would have the following changes from the original [entity OTEP](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md) to denote references within Resource:

| Field | Type | Description | Changes |
| ----- | ---- | ----------- | ------- |

> | schema_url | string | The Schema URL, if known. This is the identifier of the Schema that the entity data  is recorded in. To learn more about Schema URL ([see docs](https://opentelemetry.io/docs/specs/otel/schemas/#schema-url)) | added |

Reviewer 4: Nit: can we capitalize field names to match the style used in previous data model documents (e.g. https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#data-model)?

Author: I have them matching the proto, i.e. I expect these to be the proto changes we define.  I'm happy to change, but want to make sure we understand proposed proto changes.

Reviewer 4: OK, I am fine with either approach as long as it is consistent. If we use snake_case for entities, I think we should go and update logs and entities docs to use that too.

> | schema_url | string | The Schema URL, if known. This is the identifier of the Schema that the entity data  is recorded in. To learn more about Schema URL ([see docs](https://opentelemetry.io/docs/specs/otel/schemas/#schema-url)) | added |

Reviewer 6: I wonder if we can somehow punt on officially adding `schema_url` until we have concrete usage from stable semantic conventions. In the same way we missed the mark by prematurely adding `schema_url` to resource, we may miss the mark by adding it here if our thinking around `schema_url` evolves between now and the first stable entity semantic conventions.

Author: Understand the concern, but two things:

(1) The move to the bundle approach I think fixes the issue with current schema URL.  We can know have the *modelling* of resource attributes match the *instrumentation* of resource detection -> like we do for other semantic conventions. i.e. this better be right as I don't think we have another shot to take at it, nor do I expect us to change our minds before the first stabilized "entity/resource" hits.
(2) I'm not sure there's value in stabilizing semconv resource attribtues if we can't also communicate those versions in some way.  I'd argue not having it would block stability or make it not really something you could engage with.

Author: Let me know if you have additional concerns.  Assuming this is addressed otherwise.

| type | string | Defines the type of the entity. MUST not change during the lifetime of the entity. For example: "service" or "host". This field is required and MUST not be empty for valid entities. | unchanged |

> | identifying_attributes_keys | repeated string | Attribute Keys that identify the entity.<br/>MUST not change during the lifetime of the entity. The Id must contain at least one attribute.<br/><br/>These keys MUST exists in Resource.attributes.<br/><br/>Follows OpenTelemetry common attribute definition. SHOULD follow OpenTelemetry semantic conventions for attributes.| now a reference |

Reviewer 4: Listing just the keys complicates operations with entities (to find the value of an entity key you have to do 2 lookups instead of 1). Why not list the full key/value pairs?

Author: that's listed below in alternatives considered.

Reviewer 4: We discussed this offline. I think keeping it as key/value is more desirable. We have tools in many places that know how to deal with key/value lists. We don't have any tooling that understands key indexes.

Reviewer 8: Why not move that link completely to the entity schema? It does not make sense to ship that information with every log line that we send around in our observability stack.

> | descriptive_attributes_keys | repeated string | Descriptive (non-identifying) attribute keys of the entity.<br/>MAY change over the lifetime of the entity. MAY be empty. These attribute keys are not part of entity's identity.<br/><br/>These keys MUST exist in Resource.attributes.<br/><br/>Follows any value definition in the OpenTelemetry spec - it can be a scalar value, byte array, an array or map of values. Arbitrary deep nesting of values for arrays and maps is allowed.<br/><br/>SHOULD follow OpenTelemetry semantic conventions for attributes.| now a reference |

Reviewer 4: Why do we need descriptive_attributes_keys in the the `EntityRef`? `EntityRef` _references_ an entity. It sufficient that a reference specifies the data that _identifies_ the entity. The `Type` and `Id` fields are sufficient for identifying the entity. Why do we need to specify descriptive attributes?

Author: We have descriptive attributes in Resource today, whether we like it or not.  This allows those use cases to continue, but we can directly address and deal with them.

Reviewer 4: I think that's fine, we should keep descriptive attributes in the Resource. We don't have to have them in the EntityRefs though. Resource attributes will be a superset of EntityRef attributes.

Reviewer 5: Then is the only difference between an entity and an entity ref within resources the lack/presence of the descriptive attributes?

Author: I think we need to keep descriptive attributes so we know how to handle certain entity-merge issues.  Please see use cases listed below.

The main differences between `Entity` and `EntityRef`;

- `EntityRef` is embedded on entity.
- `EntityRef` implies a relationship between all entities on the resource (without any details on what that relationship is, besides a possible bundle/aggregation slice)
- `Entity` would be reported via an independent channel (TBD: current thinking is through events on OTLP logs).

> | descriptive_attributes_keys | repeated string | Descriptive (non-identifying) attribute keys of the entity.<br/>MAY change over the lifetime of the entity. MAY be empty. These attribute keys are not part of entity's identity.<br/><br/>These keys MUST exist in Resource.attributes.<br/><br/>Follows any value definition in the OpenTelemetry spec - it can be a scalar value, byte array, an array or map of values. Arbitrary deep nesting of values for arrays and maps is allowed.<br/><br/>SHOULD follow OpenTelemetry semantic conventions for attributes.| now a reference |

Reviewer 6: If an entity's descriptive attributes can change over the lifetime of the entity, and the descriptive attributes reference the resource attributes, then either:
- the resource attributes are now subject to change from one export to another
- OR descriptive attributes are static, but can change via entity lifecycle events

Is that how you see it?

Author: I think the nuance of allowing resource changes in an SDK will need to be in a follow up OTEP expanding on that component, likely from client-SIG.  For now - I'd like to make progress on the simple merge algorithm and move from this proposal to prototypes of the merge algorithm and show casing how it interacts with existing resource detection in various languages.

Author: Resolving as we will have a follow on OTEP to address this, as described in the OTEP.

### Problem 1: Commingling of Entities

We embrace the need for commingling entities in Resource and allow downstream users to interact with the individual entities rather than erasing these details.

### Problem 2: Lack of Precise Identity

Identity is now clearly delineated from description via the Entity portion of Resource. When Entity is used for Resource, only identifying attributes need to be interacted with to create resource identity.

### Problem 3: Lack of Mutable Attributes

This proposal offers two solutions going forward to this:

- Descriptive attributes may be mutated without violating Resource identity
- Entities whose lifetimes do not match SDK may be attached/removed from Resource.

### Problem 4: Metric Cardinality Problem

Via solution to (2) we can leverage an identity synthesized from identifying attributes on Entity.  By directly modeling entity lifetimes, we guarantee that identity changes in Resource ONLY occur when source of telemetry changes.  This solves unintended metric cardinality problems (while leaving those that are necessary to deal with, e.g. collecting metrics from phones or browser instances where intrinsic cardinality is high).

> ## Entity WG Rubric

Reviewer 4: WG = WorkGroup?

Author: yes

The Entities WG came up with a rubric to evaluate solutions based on shared
beliefs and goals for the overall effort.  Let's look at how each item is
achieved:

### Resource detectors (soon to be entity detectors) need to be composable / disjoint

Entity detection and Resource Manager now fulfill this need.

> ### New entities added by extension should not break existing code

Reviewer 4: What does "extension" refer to?

Author: I create a new semantic convention, entity detector, etc.  The existence of this new "type" should not cause breakages throughout the ecosystem.

Users will need to configure a new Entity detector for new entities being modelled.

### Navigational attributes need to exist and can be used to identify an entity but could be augmented with UUID or other aspects. - Having ONLY a UUID for entity identification is not good enough

Resource will still be composed of identifying and descriptive attributes of Entity, allowing baseline navigational attributes users already expect from resource.

### Collector augmentation / enrichment (resource, e.g.) - Should be extensible and not hard-coded. We need a general algorithm not specific rulesets

> Entity concept provides a new "bundle" mechanism to resource for the Collector to augment enrich a group of attributes and better identify conflicts (or identity changes) caused therein.

Reviewer 4: This sentence is unclear.

Author: Reworded

## Open Questions

The following remain open questions:

### How to attach Entity "bundle" information in Resource?

The protocol today requires a raw grab bag of Attributes on Resource. We cannot break this going forward.  However, Entities represent a new mechanism of "bundling" attributes on Resource and interacting with these bundles.  We do not want this to bloat the protocol, nor do we want it to cause oddities.

Going forward, we have set of options:

- Duplicate attributes in `Entity` section of Resource.
- Reference attributes of Resource in entity.
- Only identify Entity id and keep attribute<->entity association out of band.
- Extend Attribute on Resource so that we can track the entity type per Key-Value (across any attribute in OTLP).

The third option prevents generic code from interacting with Resource and Entity without understanding the model of each.  The first keeps all usage of entity simple at the expense of duplicating information and the middle is awkward to interact with from an OTLP usage perspective. The fourth is violates our stability policy for OTLP.

### How to deal with Resource/Entities whose lifecycle does not match the SDK?

> This proposal motivates a Resource Coordinator in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.

Reviewer 4: I think we don't need to broadcast anything in the TracerProvider, LogProvider, MeterProvider if we exclude descriptive attributes from the EntityRef. By definition the Type and Id of entity can't change during entity's lifetime so the EntityRef is immutable, there is no changes to broadcast.

One case that we may want to broadcast is if an entity that was previously included in a Resource via EntityRef ceases to exist. Arguably in this case we want to stop including it in the Resource. I don't know though if there is a real use case like this where the SDK outlives the Entity that was detected by SDK detector. Do we have such cases?

Author: We need to broadcast when an Entity no longer exists, at a minimum, or when a new one has replaced the previous value (e.g. Browser Session).

The main focus here is dealing with entities whose lifetimes do NOT match the SDK.

Regarding mutable/descirptive attributes let's talk more in the entities WG.  There was adamant discussion that these are needed in Resource.

Reviewer 4: >We need to broadcast when an Entity no longer exists, at a minimum, or when a new one has replaced the previous value (e.g. Browser Session).

Are we thinking about Browser Session as an entity? It does not seem to be an entity to me. Instead it seems more like an execution context and the way you would want to record this information is via Span/Metric/LogRecord attributes, not via Resource or Entity attributes. Isn't that what the [session semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/session.md) say (that it is a LogRecord attribute)?

Reviewer 4: >Regarding mutable/descirptive attributes let's talk more in the entities WG. There was adamant discussion that these are needed in Resource.

To be clear: I think it is fine to include descriptive attributes in the Resource `attributes` field _as long as_ you can ensure these attributes don't mutate during Resource lifetime. If the attributes mutate then we have a problem since Resource is assumed immutable in the SDK spec and in all current SDK implementations. I don't know how we would overcome this problem without breaking the Resource API (perhaps there is a way?). 

The reason I think it is right to include descriptive attributes in the Resource is because that's what we do today and I think we must continue doing so for backward compatibility reasons so that recipients continue seeing the same data in the Resource.

That said, I don't see why we would _also_ want to include those same descriptive attributes in the EntityRef. I don't see the purpose. If there is a purpose I would like to know.

### How to deal with Prometheus Compatibility for non-SDK telemetry?

> Today, Prometheus compatibility relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.

Reviewer 4: Do we have a doc with Prometheus compatibility rules? If yes please link to it.

A quick proposal of what this might look like:

- `target_info` metric generation is updated to exclude any keys which are contained in `descriptive_attributes_keys` of an entity.
- For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD), which has all identifying and descriptive keys. This should play nicely with the planned improvements to [info-typed metrics](https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals).

### Should entities have a domain?

> Is it worth having a `domain` in addition to type for entity?  We could force each entity to exist in one domain and leverage domain generically in resource management.  Entity Detectors would be responsible for an entire domain, selecting only ONE to apply a resource. Domains could be layered, e.g. a Cloud-specific domain may layer on top of a Kubernetes domain, where "GKE cluster entity" identifies *which* kubernetes cluster a kuberntes infra entity is part of.  This layer would be done naively, via automatic join of participating entities or explicit relationships derived from GKE specific hooks.

Reviewer 4: I think `Type` values should be namespaced like attribute values. For example Kubernetes Pods will have `Type` value of "k8s.pod". "k8s" is the namespace here, just like it is for relevant Kubernetes attributes. I think namespacing is sufficient for name conflict avoidance, we don't need an additional, separate `domain` field.

Reviewer 4: Ack and I agree with you.

>E.g. could you rely on {foo}.{other} and {foo}.{other2} meaning that other + other2 are related in some way?

We don't currently formally define this meaning in any way, but that vague meaning is there. We don't just randomly put things in namespaces, we intuitively group them by _some_ characteristic. `k8s.pod.*` attributes `k8s.node.*` attributes don't just use `k8s.` prefix by chance, it is because pod and node are both Kubernetes concepts.

I don't know if we want to have a stricter definition of what it means for the type to be in the same namespace, but I think you have a good point: we should consider it.

It's unclear if this is needed initially, and we believe this could be layered in later.

> ### Should resources have only one associated entity?

Reviewer 4: Nit: we don't define "associated" earlier. I understand that EntityRef is what we call an association. It would be useful to define this in the doc to make it clear.

Author: Agreed. I'll work on where to put that, but this should be added.

Author: Added this earlier.

> Given the problems leading to the Entities working group, and the needs of existing Resource users today, we think it is infeasible and unscalable to limit resource to only one entity.  This would place restrictions on modeling Entities that would require OpenTelemetry to be the sole source of entity definitions and hurt building an open and extensible ecosystem.  Additionally it would need careful definition of solutions for the following problems/rubrics:

Reviewer 4: You know my opinion so far is that we don't need more than one associated entity. It would be useful to list some use cases that demonstrate the need to have more than one associated entity. I would like to have a thorough debate on this topic and use cases are the necessary battleground for this debate.

Author: I believe you missed this debate in the Entities WG - but we did take [notes](https://docs.google.com/document/d/15Yt9ss2_EhuFPqItPbk4vjfpeRDAQ5WCUVuY_kCeOAo/edit#heading=h.y4e2fcmakmhd).  We can re-open this discussion in the meeting again.

I'll list the notes  here:

## 2024-07-18, 8am PT

```
Concerns around single vs. multi-associated
- Use case where eng domain was original modeled and had ONE associated entity
- Now, new domain is modeled, e.g. EC2
- How would we associate the *more* concrete type without breaking backward compatibility if systems rely on previously associated single-entity type.
```

Later
```
Entity Overlap - Should OTEL Signals have only ONE producing entity (via Resource)?
- Types of Entities
   - Infrastructure (Machine, VM, Container, CPU, Process)
   - Logical (Service, Application, etc.)
    - "Mixed" (Deployment, Session)
    
Example:
A k8s deployment leads to k8s pods being created on k8s nodes. Should OTLP metrics from this include K8S deployment identity *and* k8s pod identity *and* k8s node identity?
The user configured a service.name in opentelemetry.  It's also running on a virtual machine.  Do we include service.instance.id *and* vm identity in OTLP generated?
```

Additionally, from discussion on other design docs:

- [In the presence of "service" and "process" - which entity do you pick for resource?](https://docs.google.com/document/d/18FS9eFCW8N1oJdsKEQDJGFNVHPeZYx8rdCmRDTi9GNE/edit?disco=AAABLWnuOBw)
- [What about just the issue of picking between "service" or "container / process"?](https://docs.google.com/document/d/18FS9eFCW8N1oJdsKEQDJGFNVHPeZYx8rdCmRDTi9GNE/edit?disco=AAABLWnuOB8)
  - Note: I think the answer you give here needs a LOT of fleshing out how that solution would interact with the SDK and how users would be able to acheive the solution you lay out.  I don't see it as a viable thing to do in practice, and the design doesn't cover how to achieve it.
- A link in this OTEP on Resource, including [lessons from GCP's own single-resource model](https://docs.google.com/document/d/1Xd1JP7eNhRpdz1RIBLeA1_4UYPRJaouloAYqldCeNSc/edit#heading=h.1yfz555jem12)

And some offline discussions not recorded:

- Anna in the Entities WG had some concerns that I think weren't written down, I'll let her comment on those here.
- There's a use case where we need to allow the open ecosystem to define entity types. If a user is leveraging OpenTelemetry and all entity types *MUST* be defined within semconv this is problematic.  Take the following scenario.
  - User X is using a tool that leveraged OpenTelemetry and Semantic Conventions.  Today they interact with "process-container" for the data being generated.  Their tool understands what process-container is, and uses it for identity.
  - System Y defines a new way of modelling infrastructure.  They use containers and processes, but require a new type of identity that User X desires.
  - User X is forced to update their entity identity in Resource to be "Process-Container-System-Y".
     - They can no longer leverage OOTB entity detectionf or process or containers
     - System Y is forced to duplicate all behavior around process/container look up, even if it could be shared.
     - OpenTelemetry needs to be aware that Process-Container-System-Y and Process-Container are basically overlapping things and MAY reference the same item.
     
     
In other words, I think we've been having a thorough debate, and the consensus at the Entities WG was multiple entity associations. I think more design and outlining is needed to justify how a single entity paradigm would work in an open ecosystem, and how to address the challenges outliend.

Reviewer 8: Wouldn't entities be used to encode different abstractions like service offerings or similar? Then we would immediately have orthogonal associated entities. If such abstractions can be subject to migrations, we would even have n:m relations between the physical representation modelled in resources and the entities encoding the various orthogonal views.

Author: Services are effectively logical groupings we enforce.  I believe, already, you can have different logical groupings and therefore different associated entities with a Resource.

If look at the actual attributes added by resource detection at the bottom, you'll see I think we already have N:M relations between physical<->logical representation.  Previously, the discussion was whether have a 1:1 Resource<->Entity relationship was ideal.  Given, at a minimum, we have 1:1 logical<->physical entity relationships, and we can have multiple logical models, so N:1, We're basically loosening this view.

There's a lot of rational and reasoning in the rest of the document on this.

- New entities added by extension should not break existing code
- Collector augmentation / enrichment (resource, e.g.) - Should be extensible and not hard-coded. We need a general algorithm not specific rulesets.

### What identity should entities use (LID, UUID / GUID, or other)?

One of the largest questions in the first entities' OTEP was how to identify an entity.  This was an attempt to unify the need for Navigational attributes with the notion that only identifying attributes of Entity would show up in Resource going forward. This restriction is no longer necessary in this proposal and we should reconsider how to model identity for an Entity.  

This can be done in follow up design / OTEPs.

> ### What happens if existing Resource translation in the collector remove resource attributes an Entity relies on?

Reviewer 4: The Collector's behavior that addresses this problem is explained in a previous document [here](https://docs.google.com/document/d/1VUdBRInLEhO_0ABAoiLEssB1CQO_IcD5zDnaMEha42w/edit#heading=h.7jq82arq6q09). Any reason not to go with the suggested approach? I think it solves this problem.

Author: With the change from storing key-value pairs on Entity to just key references, this no longer works in the backwards compatible case (old collector, new data).

- Consider this a bug and warn users not to do it.
- Specify that missing attribute keys are acceptable for descriptive attribtues.
- Specify that missing attribute keys denote that entities are unusable for that batch of telemetry, and treat the content as malformed.

## Trade-offs and mitigations

The design proposed here attempts to balance non-breaking (backwards and forwards compatible) changes with the need to improve problematic issues in the Specification.  Given the inability of most SDKs to implement the current Resource merge specification, breaking this should have little effect on actual users.  Instead, the proposed merge specification should allow implementations to match current behavior and expectation, while evolving for users who engage with the new model.

## Prior art and alternatives

Previously, we have a few unaccepted oteps, e.g. ([OTEP 208](https://github.com/open-telemetry/oteps/pull/208)).  Additionally, there are some alternatives that were considered in the Entities WG and rejected.

Below is a brief discussion of some design decisions:

> - **Only associating one entity with a Resource.**  This was rejected, as too high a friction point in evolving semantic conventions and allowing independent systems to coordinate identity + entities within the OpenTelemetry ecosystem.  Eventually, this would force OpenTelemetry to model all possibly entities in the world and understand their interaction or otherwise prevent non-OpenTelemetry instrumentation from interacting with OpenTelemetry entities.

Reviewer 4: Can you clarify how does this result in a friction in evolving semantic conventions? This seems orthogonal to me, I don't understand what does it have to do with the ability to record one associated entity.

Author: From experience [having a system like this](https://cloud.google.com/monitoring/api/resources) - it lead to everyone pushing for using the same entity (thereby trying to overload its definition) or refusing to define new ones / re-using the ones "close" to what they need, because of the difficulting in reporting against different entities.

E.g. having a Collector which defines the entity/resource lookup ONCE for all producers of telemetry is a use case that becomes difficult if that collector needs different entity types by source of data.  To the point folks just re-use the one it detect by default, negating the value of Entity as you need to check attributes for identity anyway.

Additionally, having new concepts be modeled AFTER the fact is hugely disruptive and breaking of the entire ecosystem to the point we just don't.  If you realize that there is a sub-concept in the model too late in the process, it is VERY difficult to encode it and deal with it, e.g. having things associate against k8s.pod and then realizing you need k8s.container modelled and half of the telemetry you generate should change its entity type (which can break downstream users interacting with entity).

Reviewer 4: That's very useful examples, thank you.

| entities | repeated ResourceEntityRef | Set of entities that participate in this Resource. | Added |

The DataModel would ensure that attributes in Resource are produced from both the identifying and descriptive attributes of Entity.  This does not mean the protocol needs to transmit duplicate data, that design is TBD.

### ResourceEntityRef

The entityref data model, would have the following changes from the original [entity OTEP](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md) to denote references within Resource:

| Field | Type | Description | Changes |
| ----- | ---- | ----------- | ------- |
| schema_url | string | The Schema URL, if known. This is the identifier of the Schema that the entity data  is recorded in. To learn more about Schema URL ([see docs](https://opentelemetry.io/docs/specs/otel/schemas/#schema-url)) | added |
| type | string | Defines the type of the entity. MUST not change during the lifetime of the entity. For example: "service" or "host". This field is required and MUST not be empty for valid entities. | unchanged |

> | identifying_attributes_keys | repeated string | Attribute Keys that identify the entity.<br/>MUST not change during the lifetime of the entity. The Id must contain at least one attribute.<br/><br/>These keys MUST exists in Resource.attributes.<br/><br/>Follows OpenTelemetry common attribute definition. SHOULD follow OpenTelemetry semantic conventions for attributes.| now a reference |

Reviewer 8: Is adding or removing a value from the set considered a change? Do entities have versions? If so, can entities be referred to from places where the version is not known?

Author: Yes - `schema_url` can be considered a version for telemetry within OTel -> this would be the version of the identity.

Yes - we likely need to deal with the possibility folks do not specify a schema_url for an entity. In this case, it would be an unknown version and we are unable to determine if we can merge.  I'm updating the merge algorithm definition shortly, so I'll account for that in the update.

Reviewer 8: I wasn't referring to schema versions. I was thinking about something like `service.version` for entities to signal the version of the content, not the type.

Author: How would the version not be part of the entity identity in that case?

Reviewer 8: OK, I spent some time reading through all the documents and some referenced documents. I like the overall approach, but fear that the lack of clarity might be an issue. Would it make sense to ask my organization if I could contribute to such projects and somehowe become a contributing member of this work stream? Something that I really do not get (and that's maybe the source of my confusion) is why we would send the attribute names twice. If the proposal envisions separate entity state events (which is something I favor), it would make sense to reduce the data sent with data points as much as possible. Also, not sending the attribute keys here would completely remove situations where the set of identifying attributes varies. Furthermore, the MUST not change statements here are not correct. They CANNOT change as they define equivalence of the underlying entity and changing them must result in a separate entity instance. Finally, I still struggle with partial identification: The service entity, for compatibility reasons, would be identified by `{service.name, service.namespace, service.instance.id}`. However, in practice, there are cases where e.g. `service.instance.id` isn't known when producing the data. It might get added later by a processor or it might stay like that. Using entities might fix that, but the transition towards entities would result in a phase where this issue exists and must be dealt with.

Author: Sorry for not responding to this earlier.

IF you're interested in contributing, we are trying to organize our prototyping work.  You can join/discuss with us live (or watch recordings of) the [Specification: Entities WG meeting](https://github.com/open-telemetry/community/?tab=readme-ov-file#specification-sigs).

Regarding sending keys twice:  The rationale is in the document, but to explicitly state it:

- We cannot break existing `Resource` usage, which means the `attributes` fields MUST be usable without understanding or interacting with the new entity model. This means *all* attributes from entities MUST be in this portion of the proto or we've broken open telemetry.
- We debated between the following options (called out in options considered):
   - Duplicate attributes in `Resource.attributes` and `Resource.entities`.
   - Reference `Resource.attributes` in `Resource.entitites` by index
   - Reference `Resource.attributes` in `Resource.entities` by name
   
   We chose the last option for the proposal, but this may change as we prototype and move forward.  The first requirement will not.

Regarding an `entity` where identity is not known when adding to resource: Going forward we're basically outlawing that use case.  For service this means, you'll need to attach service name and instance when attaching service.  We have done work in semconv and the specification such that we can attach service.instance.id when producing data.  Additioanlly, if you want a different instance id you can do one of two things:

1. Update instance id in the collector via an entity-override
2. Create an association between the "service reported by the sdk" and the "service you get identity from somewhere else".

Reviewer 8: I'll try to join the WG on behalf of my organization or my employer.

Until then: Could we consider not setting any reference in the gRPC messages and having them behind the schema_url as a fourth option at least? I agree that not breaking existing `attributes` and having entities as transparent extension are non-negotiable requirements. Regarding the instance.id discussion: I think it makes sense to explain why I'm raising this example once I figured under which mandate I can join the WG.

Reviewer 8: (I think we can resolve this discussion by adding the option because the other parts should not block progress)

Author: The mechanism of how to understand entity semantics in Resource is an open question below, meaning while this OTEP proposing something, we aren't committed until we have prototypes we're comfortable with and these would enter the specification.

While I discuss in a different comment why using the contents of schema_url is off the table for the general algorithm, it CAN be used when we have it. Look forward to the ideas you have here.

Marking this resolve, hope you're able to join.

## Future Posibilities

This proposal opens the door for addressing issues where an Entity's lifetime does not match an SDK's lifetime, in addition to providing a data model where mutable (descriptive) attributes can be changed over the lifetime of a resource without affecting its idnetity.  We expect a follow-on OTEP which directly handles this issue.

## Use Cases

Below are a set of use cases to help motivate this design.

### SDK - Multiple Detectors of the same Entity type

Let's consider the interaction of the SDK in the presence of multiple registered
entity detectors:

```mermaid
flowchart LR
    SDK["`**SDK**`"] -->|OTLP| BACKEND["`**Backend**`"]
    SDK -.- RC((Resource Coordinator))
    RC -.- OTEL_DETECTOR((OpenTelemetry Default Resource Detection))
    RC -.- GCP_DETECTOR((Google Cloud Specific Resource Detection))
    GCP_DETECTOR -. Detects .-> GCE{{gcp.gce}}

>     GCP_DETECTOR -. Detects .-> GCPHOST{{"host (gcp)"}}

Reviewer 7: nit: diagram is broken

Author: I'm not seeing the breakage on my view, but it certainly WAS broken before [f679af2](https://github.com/open-telemetry/oteps/pull/264/commits/f679af21d880947c715f1e701e90bfe930805d07).  Please reopen if it's still broken.

  - *(new) Managing Entity changes during SDK lifetime, specifically dealing with entities that have lifetimes shorter than the SDK*

#### Resource Coordinator

The SDK Resource Coordinator is responsible for running all configured Resource and Entity Detectors.  There will be some (user-controlled, otel default) priority order to these.

- The Resource Coordinator will detect conflicts in Entity of the same type being discovered and choose one to use.
- When using Entity Detectors and Resource detectors together, the following merge rules will be used:
  - Entity merging will occur first resulting in an "Entity Merged" Resource (See [algorithm here](#entity-merging-and-resource)).
  - Resource detectors otherwise follow existing merge semantics.
    - The Specification merge rules will be updated to account for violations prevalent in ALL implementation of resource detection.
    - Specifically: This means the [rules around merging Resource across schema-url will be dropped](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge).  Instead only conflicting attributes will be dropped.

>     - SchemaURL on Resource will need to be deprecated with entity-specific schema-url replacing it. SDKs will no longer fill out SchemaURL on Resource. Additionally, as no (non-service) Resource semantic conventions have ever stabilized, SchemaURL usage on Resource cannot be in stable components of OpenTelemetry.  Given prevalent concerns of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal.

Reviewer 4: I think SchemaURL field is part of Stable specification document: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md, which means this is a breaking change, I don't think we are allowed that.

Is there a way we can continue including SchemaURL on the Resource? It does not need to perfect (it is not perfect today). For example, I think we can do this: 

If all Entities use the same SchemaURL we use that value for Resource's SchemaURL too. If there are any difference in SchemaURL values of Entities than we leave Resource's SchemaURL value empty. This will be no worse than the currently specified behavior of Resource merging where attempting to merge with different SchemaURLs is [considered an error](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge).

I think this behavior is backward compatible and we are not breaking the "Stable" document promise.

Reviewer 8: Wouldn't marking it for deprecation, but not removing it equally not break the "stable" document promise? It would cause it to remain until some v2.0.0 API is released, but the intention would be expressed clearly.

Author: I updated this description.

As  a note though, the ONLY resources to have stabilized are Service + Telemetry SDK, so any other usage of SchemaURL within open-telemetry is NOT stable, and therefore not a breaking change, according to the rules we've set forth in our documents.

Author: Marking this as resolved, please reopen if you see continued issues with how the SchemaURL section is phrased.

  - An OOTB ["Env Variable Entity Detector"](#environment-variable-detector) will be specified and provided vs. requiring SDK wide ENV variables for resource detection.
- *Additionally, Resource Coordinator would be responsible for understanding Entity lifecycle events, for Entities whose lifetimes do not match or exceed the SDK's own lifetime (e.g. browser session).*

#### Entity Detector

> The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK.  For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod.   SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

Reviewer 8: I think the document introduces the concept of descriptive and identifying attributes and it is done here. As this is a key concept, two more sentences should be spent here explaining the general concept *and their impact on equivalence on attribute sets*.

Author: Agreed - We've discussed those concepts a bunch in the past, but they're not formalized. I'll look into how to better outline them in other sections.

Reviewer 4: The descriptive and identifying attributes were introduced in [OTEP0256](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md). See the table in "Data Model" section.

Author: Given previous otep is (now) linked immediately in this document, I'm closing this as resolved.

An Entity Detector would have an API similar to:

```rust
trait EntityDetector
  pub fn detect_entities(...) -> Result<Vec<Entity>, EntityDetectionError>>
```

> Where `Result` is the equivalent of error channel in the language of choice (e.g. in Go this would be `entities, err := e.detectEntities()`).

Reviewer 8: This comment is caused by using an obscure language optimized for hard realtime programming. In Java, the result would simply be an ArrayList of entities and throw some DetectionException where applicable. The API proposal would be much clearer if any language with exceptions were chosen.

Author: We're not picking one language over another, and yes -> in java you can throw exceptions.

Reviewer 8: That's not my point. It could also be written in some pseudo code with exceptions letting exception-averse languages pick their favorite error propagation mechanism. The point is, that the statement here is only required because one of these mechanisms is chosen over the more general concept.

Author: This is the second such comment about this code, the first asking for a go example (which is one I added), now you're asking for an exception throwing example.

I'm not picking exception vs. non-exception throwing languages here. The code is an example interface, you haven't shown me that this makes it difficult to understand or interpret, given your suggestion shows exactly how to translate into exception throwing languages.

I think the main point here is to *expliclty* show that the whole set of entities is returned or an error, which is important to make for non-exception languages where that isn't an implciit assumption.  I would do this even in pseudocode.

Reviewer 8: Maybe we should change the example to pseudo code to avoid wasting time on programming language discussions.

Author: I'm closing this as resolved - won't fix.  If there's an aspect of the design that is struggling to be communicated, I'll fix it.  The comment implies that's not the case.

#### Entity Merging and Resource

The most important aspect of this design is how Entities will be merged to construct a Resource.

We provide a simple algorithm for this behavior:

- Construct a set of detected entities, `E`

>   - All entity detectors are sorted by priority (highest first)

Reviewer 8: The EntityDetector API does not expose an explicit order. This reads like an implicit "in order of the returned sequence". Hence, it should also be phrased like that and the detector API should explain that the result is an ordered sequence affecting priority in conflict resolution.

Author: The API does not, the configuration of the Resource Manager does - and the notion of having an order is explciitly called out.

Reviewer 8: It's a single statement below that I must have read and ignored. I'd fix that by providing a forward reference to that section replacing the (highest first) comment. The order description below in the rource manager description should explain that if it is necessary at all. To me, the overall algorithm reads like `for(var D : ResourceManager.entityDetectors()) {....}` (which looks horrible with the D :) )

  - For each entity detector `D`, detect entities
    - For each entity detected, `d'`
      - If an entity `e'` exists in `E` with same entity type as `d'`, do one of the following:

>         - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:

Reviewer 8: are omitted `schema_url`s identical? or do you force consumers into using schema_urls?

Author: omitted schema_urls are considered different.   Basically imagine a non-existent schema-url as some random ID that is disjoint from all possible other schema urls.

Reviewer 8: I can accept both choices and have seen both in the past. Can we add a note on that to the section maybe citing semantic of schema_urls?

Author: If you haven't, I'd read the original resource merge specification linked to above - where that semantic/precedence is set.

Reviewer 8: If you refer to (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge), I've read it and it is, TBH, a required read to understand what "merging" is in this context. Nonetheless, the linked document isn't very specific. If you consider that the semantics of a schema_url, then I'd kindly ask you to add what you explained to the comment.

Author: this is a proposal against the otel specification.  The specification is required reading for understanding the context of the proposal.

If the linked document wasn't specific - it's an issue in the specification we need to improve.

Part of the art here though is NOT being specific.  Being too prescriptive can lead to an unimplementable specification, or one too tied to specific languages and unable to be used broadly.

          - For each descriptive attribute `da'` in `d'`
            - If `da'.key` does not exist in `e'`, then add `da'` to `ei`

>             - otherwise, ignore.

Reviewer 8: implementations should have an option to debug this case; in most cases, this should be an error

Author: Definitely disagree. In multi-observer case or even the use case of AWS Ec2 providing host entity, it's not guaranteed every observer *can* see all possible attributes.

Reviewer 8: Ok, but then you disagree with the *most* cases part of my statement, right? I would hope to be able to design our observability in a way that every entity has a defined point of enrichment on the path between source and first store.

Author: yes "most" is what I disagree with.

Your goal for how you want to design o11y -> I think this is an ideal we should aim for, but we I don't think it's something we can force.

Reviewer 8: That's why I requested to have an option ;)

Author: The design leaves available the option.

I think every scenario in this manager eneds to be debuggable.

        - If the entity identity is the same, but schema_url is different: drop the new entity `d'`

>           *Note: We could offer configuration in this case*

Reviewer 8: Do you envision entity schemas with comparable versions? If so, wouldn't we expect the highest version to win?

Author: We have not specified comparable versions yet, only equality comparison.

YES I think we should work towards comparable versions, but this is written against the spec we have now. Comparable schema URL would be a different OTEP.

Reviewer 8: In that case, I'd simply add a note, that this is the case and the change in outcome when switching is anticipated and considered acceptable.

Author: It's already in the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/README.md#schema-version-number

When I talk about comparable schema URL - I was referencing *compatibility* (i.e understanding breaking changes).  We can ensure highest wins, however that is not the algorithm I've defined here.

This algorithm is naive and consistent.  Without understanding *compatibility* of version (just ordering), we always prefer the first entity we see in the priority list of entity detectors.  If we see more than one entity we cannot assume the version is important, just the order users defined.   This is the behavior they have today and what we'll preserve.

Reviewer 8: @jsuereth there should be a statement, that, at the moment, it is a goal to provide only solutions that do not require looking into schema_urls. That would be part of a "key assumptions" section close to the beginning of the document. However, I'm not sure if making such a change in the current stage is a good idea as it would result in a restructuring of the entire document.

An Entity Detector would have an API similar to:

```rust
trait EntityDetector
  pub fn detect_entities(...) -> Result<Vec<Entity>, EntityDetectionError>>
```

Where `Result` is the equivalent of error channel in the language of choice (e.g. in Go this would be `entities, err := e.detectEntities()`).

An Entity Detector MUST NOT provide two entities of the same entity type (e.g. two `host` or two `service` entities).

#### Entity Merging and Resource

The most important aspect of this design is how Entities will be merged to construct a Resource.

We provide a simple algorithm for this behavior:

- Construct a set of detected entities, `E`
  - All entity detectors are sorted by priority (highest first)
  - For each entity detector `D`, detect entities
    - For each entity detected, `d'`
      - If an entity `e'` exists in `E` with same entity type as `d'`, do one of the following:

>         - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:

Reviewer 9: Please clarify this sentence:

> - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:

Who's "entity" in _If the entity..._? I can interpret this as:

``` python
if e.identity == e.schema_url:
	merge_first_into_second(d, e)
```

And we probably don't want to compare different attributes of the same entity, right?

Do entities have a `schema_url` attribute? From [this](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#data-model) I was under the impression an entity only attributes were `type`, `id` and `attributes`.

Also, isn't `identity` a `map<string, attribute>`? How can we compare it to `schema_url` which is a string?

Author: > Do entities have a schema_url attribute? From [this](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#data-model) I was under the impression an entity only attributes were type, id and attributes.

Yes entities have a schema_url.

> Also, isn't identity a map<string, attribute>? How can we compare it to schema_url which is a string?

Identity is a `KeyValueList`.  You compare identity *and* you compare schema_url.  I.e. you check if:

`lhs.schema_url == rhs.schema_url && lhs.identifying_attributes == rhs.identifying_attributes`

Reviewer 8: I think it is somewhat confusing that schema_*url* is treated as an opaque value consistently here. It seems that there is a silent assumption that it would be a bad idea to actually fetch the content of the URL and buffer it locally.

Author: Yes: This was explicitly talked about but didn't make it into the proposal.  We need to be able to execute our algorithm in environments where there is no access to public internet, and users may not have silently provided internal access to these URLs.

Therefore while we can *allow* it to be fetched and buffered (and we plan to use it in such a way) the core algorithm must work without this.

Author: I added a section to call out this assumption - see the "Why don't we download schema url contents?" section

Reviewer 8: @jsuereth thanks; Imho that section resolves the issue in the current state of the elaboration.

    - SchemaURL on Resource will be deprecated with entity-specific schema-url replacing it. SDKs will only fill out SchemaURL on Resource when SchemaURL matches across all entities discovered. Additionally, only existing stable resource attributes can be used in Resource SchemaURL in stable OpenTelemetry components (Specifially `service.*` and `sdk.*` are the only stabilized resource convnetions). Given prevalent concerns of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal, and existing usage was within the "experimental" phase of semantic conventions.
  - An OOTB ["Env Variable Entity Detector"](#environment-variable-detector) will be specified and provided vs. requiring SDK wide ENV variables for resource detection.
- *Additionally, Resource Provider would be responsible for understanding Entity lifecycle events, for Entities whose lifetimes do not match or exceed the SDK's own lifetime (e.g. browser session).*

#### Entity Detector

> The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod. SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

Reviewer 8: From my practical experience, I doubt that it *can* be more specific and meaningful.

> The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod. SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

Author: This is use case based.  There are some descriptive attributes that are valuable to group/slice by in resource used today (see the list of resource attributes listed below which are not identifying).

This is aimed to not be too prescriptive so we can preserve existing resource behavior.  I think for semantic conventions we'll want to provide a "bar" on when to do this, but for general specification and users, we should leave it open.

> The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod. SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

Reviewer 10: Would it make sense to further distinguish 'required' attributes per entity type which are then always included in descriptive attributes? Otherwise to determine which attributes to expect one would have to 'read the memo' per SDK?

> The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod. SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

Author: Yes - but this would be done in semantic conventions, similar to what we do for metrics/spans/logs: see [http spans](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-client) for example.

I.e. I think the "requirement level" is ok to be defined separate from the data itself.  The notion of identity + descriptive should be encoded in the transmitted data - i.e. we assume all required data is already on the entity when we get it.

> The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod. SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.

Reviewer 8: I doubt that requirements levels have any practical relevance in the sense of a software contract. Equally, it is not plausible that a descriptive attribute is required. "desired" would be a better word. But that's a digression and the discussion should be resolved here.

An Entity Detector MUST NOT provide two entities of the same entity type (e.g. two `host` or two `service` entities).

#### Entity Merging and Resource

The most important aspect of this design is how Entities will be merged to construct a Resource.

We provide a simple algorithm for this behavior:

- Construct a set of detected entities, `E`
  - All entity detectors are sorted by priority (highest first)
  - For each entity detector `D`, detect entities
    - For each entity detected, `d'`
      - If an entity `e'` exists in `E` with same entity type as `d'`, do one of the following:
        - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:
          - For each descriptive attribute `da'` in `d'`
            - If `da'.key` does not exist in `e'`, then add `da'` to `ei`
            - otherwise, ignore.
        - If the entity identity is the same, but schema_url is different: drop the new entity `d'`
          *Note: We could offer configuration in this case*
        - If the entity identity is different: drop the new entity `d'`.
      - Otherwise, add the entity `d'` to set `E`
- Construct a Resource from the set `E`.

>   - If all entities within `E` have the same `schema_url`, se the resource's

Reviewer 11: That looks like a typo?
```suggestion
  - If all entities within `E` have the same `schema_url`, set the resource's
```

## How this proposal solves the problems that motivated it

Let's look at some motivating problems from the [Entities Proposal](https://docs.google.com/document/d/1VUdBRInLEhO_0ABAoiLEssB1CQO_IcD5zDnaMEha42w/edit#heading=h.atg5m85uw9w8):

### Problem 1: Commingling of Entities

We embrace the need for commingling entities in Resource and allow downstream users to interact with the individual entities rather than erasing these details.

### Problem 2: Lack of Precise Identity

Identity is now clearly delineated from description via the Entity portion of Resource. When Entity is used for Resource, only identifying attributes need to be interacted with to create resource identity.

### Problem 3: Lack of Mutable Attributes

This proposal offers two solutions going forward to this:

> - Descriptive attributes may be mutated without violating Resource identity

Reviewer 8: This is essentially a consequence of how identity on entities is defined.

> - Descriptive attributes may be mutated without violating Resource identity

Author: Added a section detailing Resource identity, under the data model.

> - Descriptive attributes may be mutated without violating Resource identity

Reviewer 10: Makes sense, thank you

Today, [Prometheus compatibility](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md) relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.

A quick proposal of what this might look like:

- `target_info` metric generation is updated to exclude any keys which are contained in `descriptive_attributes_keys` of an entity.
- For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD), which has all identifying and descriptive keys. This should play nicely with the planned improvements to [info-typed metrics](https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals).

### Should entities have a domain?

> Is it worth having a `domain` in addition to type for entity?  We could force each entity to exist in one domain and leverage domain generically in resource management.  Entity Detectors would be responsible for an entire domain, selecting only ONE to apply a resource. Domains could be layered, e.g. a Cloud-specific domain may layer on top of a Kubernetes domain, where "GKE cluster entity" identifies *which* kubernetes cluster a kuberntes infra entity is part of.  This layer would be done naively, via automatic join of participating entities or explicit relationships derived from GKE specific hooks.

Author: I like this suggestion.  For the purpose of this OTEP - I think that's follow on work, hence leaving this as an open question.

If you'd like to take a gander at semantic convention resources (which already, mostly, have namespaces) and map this to your suggestion, I think some entities are "flat"/"global" like `host` or `container`.

Regarding registration -> We want to avoid central registration as a requirement.   I'd prefer if we do namespace we leave some convention where non-otel namespaces can have guaranteed ways not to conflict with each other via convention in naming of these namespaces.

### What identity should entities use (LID, UUID / GUID, or other)?

One of the largest questions in the first entities' OTEP was how to identify an entity.  This was an attempt to unify the need for Navigational attributes with the notion that only identifying attributes of Entity would show up in Resource going forward. This restriction is no longer necessary in this proposal and we should reconsider how to model identity for an Entity.  

This can be done in follow up design / OTEPs.

### What happens if existing Resource translation in the collector remove resource attributes an Entity relies on?

While we expect the collector to be the first component to start engaging with Entities in an architecture, this could lead to data model violations.  We have a few options to deal with this issue:

- Consider this a bug and warn users not to do it.
- Specify that missing attribute keys are acceptable for descriptive attribtues.
- Specify that missing attribute keys denote that entities are unusable for that batch of telemetry, and treat the content as malformed.

### What about advanced entity interaction in the Collector?

> On problem that motivated this design is the issue of "local resource detection" vs. "remote signal collection" in the OpenTelemetry collector.  That is, I have a process running on a machine writing to an OpenTelemetry

Reviewer 12: ```suggestion
One problem that motivated this design is the issue of "local resource detection" vs. "remote signal collection" in the OpenTelemetry collector.  That is, I have a process running on a machine writing to an OpenTelemetry
```

> collector running on a different machine.  The current `resourcedetectoinprocessor` in the collector appends attributes to resource based on discovering *where the collector is running*.  However,

Reviewer 12: ```suggestion
collector running on a different machine.  The current `resourcedetectionprocessor` in the collector appends attributes to resource based on discovering *where the collector is running*.  However,
```

## Prior art and alternatives

Previously, we have a few unaccepted oteps, e.g. ([OTEP 208](https://github.com/open-telemetry/oteps/pull/208)).  Additionally, there are some alternatives that were considered in the Entities WG and rejected.

Below is a brief discussion of some design decisions:

- **Only associating one entity with a Resource.**  This was rejected, as too high a friction point in evolving semantic conventions and allowing independent systems to coordinate identity + entities within the OpenTelemetry ecosystem.  Eventually, this would force OpenTelemetry to model all possibly entities in the world and understand their interaction or otherwise prevent non-OpenTelemetry instrumentation from interacting with OpenTelemetry entities.
- **Embed fully Entity in Resource.** This was rejected because it makes it easy/trivial for Resource attributes and Entities to diverge.  This would prevent the backwards/forwards compatibility goals and also require all participating OTLP users to leverage entities. Entity should be an opt-in / additional feature that may or may not be engaged with, depending on user need.

> - **Re-using resource detectoin as-is** This was reject as not having a viable compatibility path forward.  Creating a new set of components that can preserve existing behavior while allowing users to adopt the new functionality means that users have better control of when they see / change system behavior, and adoption is more obvious across the ecosystem.

Reviewer 12: ```suggestion
- **Re-using resource detection as-is** This was rejected as not having a viable compatibility path forward.  Creating a new set of components that can preserve existing behavior while allowing users to adopt the new functionality means that users have better control of when they see / change system behavior, and adoption is more obvious across the ecosystem.
```

## Future Posibilities

> This proposal opens the door for addressing issues where an Entity's lifetime does not match an SDK's lifetime, in addition to providing a data model where mutable (descriptive) attributes can be changed over the lifetime of a resource without affecting its idnetity.  We expect a follow-on OTEP which directly handles this issue.

Reviewer 12: ```suggestion
This proposal opens the door for addressing issues where an Entity's lifetime does not match an SDK's lifetime, in addition to providing a data model where mutable (descriptive) attributes can be changed over the lifetime of a resource without affecting its identity.  We expect a follow-on OTEP which directly handles this issue.
```

## Use Cases

Below are a set of use cases to help motivate this design.

### SDK - Multiple Detectors of the same Entity type

Let's consider the interaction of the SDK in the presence of multiple registered
entity detectors:

```mermaid
flowchart LR
    SDK["`**SDK**`"] -->|OTLP| BACKEND["`**Backend**`"]
    SDK -.- RC((Resource Provider))
    RC -.- OTEL_DETECTOR((OpenTelemetry Default Resource Detection))
    RC -.- GCP_DETECTOR((Google Cloud Specific Resource Detection))
    GCP_DETECTOR -. Detects .-> GCE{{gcp.gce}}
    GCP_DETECTOR -. Detects .-> GCPHOST{{"host (gcp)"}}
    OTEL_DETECTOR -. Detects .-> HOST{{"host (generic)"}}
    OTEL_DETECTOR -. Detects .-> PROCESS{{process}}
    OTEL_DETECTOR -. Detects .-> SERVICE{{service}}
```

> Here, there is a services running on the Google Compute Engine. The user

Reviewer 12: ```suggestion
Here, there is a service running on Google Compute Engine. The user
```

Reviewer 12: Drop the "the." It's cleaner. :)

has configured a Google Cloud specific set of entity detectors.  Both the
built in OpenTelemetry detection and the configured Google Cloud detection
discover a `host` entity.

The following outcome would occur:

- The resulting resource would have all of the following entities: `host`, `process`, `service` and `gcp.gce`
- The user-configured resource detector would take priority over built in: the `host` defined from the Google Cloud detection would "win" and be included in resource.

>   - This means `host.id` e.g. could be the id discovered for GCE VMs.  Similarly for other cloud provider detection, like Amazon EC2 where VMs are given a unique ID by the Cloud Provider, rather than a generic machinne ID, e.g.

Reviewer 12: ```suggestion
  - This means `host.id` e.g. could be the id discovered for GCE VMs.  Similarly for other cloud provider detection, like Amazon EC2 where VMs are given a unique ID by the Cloud Provider, rather than a generic machine ID, e.g.
```

  - This matches existing behavior/expectations today for AWS, GCP, etc. on what `host.id` would mean.
- Users would be able to configure which host wins, by swapping the priority order of "default" vs. cloud-specific detection.

### SDK and Collector - Simple coordination

Let's consider the interaction of resource, entity in the presence of an SDK
and a Collector:

```mermaid
flowchart LR
    SDK["`**SDK**`"] -->|OTLP| COLLECTOR["`**Collector**`"]
    COLLECTOR -->|OTLP| BACKEND["`**Backend**`"]
    SDK -.- RC((Resource Provider))
    COLLECTOR -.- RP((Resource Processor))
    RP -. Detects .-> EC2{{aws.ec2}}
    RP -. Detects .-> HOST{{host}}
    RC -. Detects .-> PROCESS{{process}}
    RC -. Detects .-> SERVICE{{service}}
```

Here, an SDK is running on Amazon EC2. it is configured with resource detection

> that find a `process` and `service` entity.  The SDK is sending data to an

Reviewer 12: ```suggestion
that finds a `process` and `service` entity.  The SDK is sending data to an
```

OpenTelemetry Collector that has a resource processor configured to detect
the `ec2` and `host` entities.

The resulting OTLP from the collector would contain a resource with all
of the entities (`process`, `service`, `ec2`, and `host`).  This is because
the entities are all disjoint.

*Note: this matches today's behavior of existing resource detection and OpenTelmetry collector where all attributes wind up on resource.*

### SDK and Collector - Entity coordination with descriptive attributes

Let's consider the interaction of resource, entity where both the SDK and the Collector detect an entity:

```mermaid
flowchart LR
    SDK["`**SDK**`"] -->|OTLP| COLLECTOR["`**Collector**`"]
    COLLECTOR -->|OTLP| BACKEND["`**Backend**`"]
    SDK -.- RC((Resource Provider))
    COLLECTOR -.- RP((Resource Processor))
    RP -. Detects .-> HOST2{{host}}
    RC -. Detects .-> HOST{{host}}
    RC -. Detects .-> SERVICE{{service}}
```

> Here, and SDK is running on a machine (physical or virtual).  The SDK is

Reviewer 12: ```suggestion
Here, an SDK is running on a machine (physical or virtual).  The SDK is
```

- `k8s.pod.*` - from otel operator
- `k8s.container.*` - from otel operator
- `k8s.cluster.*` - from EKS detector on SDK or collector
- `cloud.*` - from EKS detector on SDK or collector

A simple litmus test derived from this for when to include an entity on
Resource would be: "Any entity relevant to the produced telemetry should be
included".

However, this can be refined.  Resources today provide a [few key features](https://docs.google.com/document/d/1Xd1JP7eNhRpdz1RIBLeA1_4UYPRJaouloAYqldCeNSc/edit):

- They provide identity - Uniquely identifying the origin of the data.
- They provide "navigationality" - allowing users to find the source of the data within their o11y and infrastructure tools.
- They allow aggregation / slicing of data on interesting domains.

A litmus test for what entities to include on resource should be as follows:

- Is the entity the source/origin of the data?
- Does the entity help navigate to the source of the data? (e.g. `k8s.cluster.*` helping find a `k8s.container.*`)
- Do want to easily slice/aggregate on an axis provided by the entity? (e.g. quickly filtering all CPU container usage metrics across a cluster to find overloaded nodes).

> If the answer to any question is yes, then include the entity on resource.

Reviewer 10: This is quite open-ended. Do you think that limiting a single domain to only include a single most specific detected entity in the resource is an option? And everything else is reported via relations instead of bundling everything together. That way the above example could be reduced to the following entities:

- `service`
- `telemetry.sdk`
- `process`
- `host`
- `container`
- `k8s.pod`
- `cloud.?`

This assumes that we can either somehow distinguish non-prefixed types into relevant domains, i.e. based on 'Schema Family identifier' of the schema_url (to preserve backward compatibility) or we would need to prefix every entity type (i.e. host -> infra.host) for consistency

Author: It depends on how we define identity.  With how things work today - We *cannot* limit to just one per domain.  We would need to alter entities in semantic conventions.  E.g. `k8s` related attributes today require *many* entities to really be useful.

E.g.  `k8s.namespace.name`, `k8s.pod.name`, `k8s.cluster.name` <- these are across three "entities" today.

If you have time - I'd like to see a proposal on how to address this with domains to help us answer that open question.  It needs to address what we do about semconv today and how evolve the model.

For the purpose of this proposal - I think we need to spend time addressing domains and entity modelling, but I believe we can do so while prototyping - Hence this remains an open question for us to resolve as we move forwards.

Going forward, we have set of options:

- Duplicate attributes in `Entity` section of Resource.
- Reference attributes of Resource in entity.
- Only identify Entity id and keep attribute<->entity association out of band.
- Extend Attribute on Resource so that we can track the entity type per Key-Value (across any attribute in OTLP).

The third option prevents generic code from interacting with Resource and Entity without understanding the model of each.  The first keeps all usage of entity simple at the expense of duplicating information and the middle is awkward to interact with from an OTLP usage perspective. The fourth is violates our stability policy for OTLP.

### How to deal with Resource/Entities whose lifecycle does not match the SDK?

This proposal motivates a Resource Provider in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.

### How to deal with Prometheus Compatibility for non-SDK telemetry?

> Today, [Prometheus compatibility](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md) relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.

Reviewer 13: @jsuereth could you here mention the possible solution of synthesizing a single identifier, for use within Prometheus, from the set of identifying entity attributes? E.g. it would become the `target_info` `instance` label when converting OTel metrics to Prometheus format. You mentioned in Slack we could use a "consistent-hash".

Author: Absolutely!

Author: done.

A quick proposal of what this might look like:

- `target_info` metric generation is updated to exclude any keys which are contained in `descriptive_attributes_keys` of an entity.

> - For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD), which has all identifying and descriptive keys. This should play nicely with the planned improvements to [info-typed metrics](https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals).

Reviewer 13: What is the foreseen benefit of excluding keys from `descriptive_attributes_keys` of an entity, when generating `target_info`? I'm not able to connect this to the goal of "outlining identity in all resources within OpenTelemetry".

Author: I took @dashpole's suggestion here. I think in following up, we may instead go with an approach of synthesizing job/instance labels in ALL cases for prometheus, so we can leverage `*_info` metrics.  I believe the goals as stated were:

1. Support ingestion + export (OTLP->Prometheus->OTLP).
2. Avoid breaking existing usage (that has `target_info`).

Maybe for this we should just list *requirements* for a prometheus compatibility design and let ourselves find solutions over time, the same as the ENV variable section does.  WDYT?

Author: I took a first crack at defining some requirements, PTAL

Reviewer 13: > For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD)

Isn't this really the same as the option in the paragraph below ("that would break compatibility")? I wonder if this paragraph instead should mention synthesizing a `target_info` `instance` label from all the identifying resource attributes.

Author: Will do - let's discuss tomorrow and I'll update options based on that.

Author: Options are updated in preparation for our discussion, PTAL

      - process.command_args
      - process.owner
    - [builtin](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/resource/builtin.go)
      - service.instance.id
      - service.name
- [OTEL operator](https://github.com/open-telemetry/opentelemetry-operator/blob/a1e8f927909b81eb368c0483940e0b90d7fdb057/pkg/instrumentation/sdk_test.go#L752) injected ENV variables
  - service.instance.id
  - service.name
  - service.version
  - k8s.namespace.name
  - k8s.pod.name
  - k8s.node.name
  - k8s.container.name

### Implications

Some initial thoughts on implications:

AWS, Azure, GCP, Heroku, etc. all provide the following "bundles" of resource:

- `cloud.*`
- `faas.*`, when relevant
- `host.*`, when relevant
- `k8s.cluster.*`, when relevant
- `service.*` when relevant
- `container.*` for a subset of k8s providers

"system" detection provides the following:

- `host.*`
- `os.*`
- `process.*` for SDKs
- `container.*` for Docker images

SDK specific detection provides the following:

- `sdk.*`
- `service.*`

The OTEL operator for k8s provides the following via ENV variables:

- `k8s.namespace.*`
- `k8s.node.*`
- `k8s.pod.*`
- `k8s.container.*`
- `service.*`

> ### What could this mean for chosing entities that belong on resource?

Reviewer 13: ```suggestion
### What could this mean for choosing entities that belong on resource?
```

An Entity Detector MUST NOT provide two entities of the same entity type (e.g. two `host` or two `service` entities).

#### Entity Merging and Resource

The most important aspect of this design is how Entities will be merged to construct a Resource.

We provide a simple algorithm for this behavior:

- Construct a set of detected entities, `E`
  - All entity detectors are sorted by priority (highest first)
  - For each entity detector `D`, detect entities
    - For each entity detected, `d'`
      - If an entity `e'` exists in `E` with same entity type as `d'`, do one of the following:
        - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:
          - For each descriptive attribute `da'` in `d'`
            - If `da'.key` does not exist in `e'`, then add `da'` to `ei`
            - otherwise, ignore.
        - If the entity identity is the same, but schema_url is different: drop the new entity `d'`
          *Note: We could offer configuration in this case*
        - If the entity identity is different: drop the new entity `d'`.
      - Otherwise, add the entity `d'` to set `E`
- Construct a Resource from the set `E`.

>   - If all entities within `E` have the same `schema_url`, set the resource's

Reviewer 9: What about the resource attributes? Are they also dependent on the entities set `E`?

Author: Yes.

You construct a Resource using this entity set.

- The `Resource.attributes` field contains *alL* identifying and descriptive attributes from the entity set.
- The `Resource.schema_url` field is set to the schema_url shared by all entities (if they all have the same one) or unset if they do not have the same schema_url.
- The `Resource.entities` field would contain the EntityRef proto, which just contains key names and the entity type for the entity set.

Reviewer 9: What is an `EntityRef` proto?

Author: It's defined later in the document.  We'll be creating that and contributing it to OTLP as part of this OTEP.

```yaml
processor:
  entityresourcedetection:
     # Order determines override behavior
     detectors: [env, docker]
     # False means only append if entity doesn't already exist.
     override: false 
```

The list of detectors is given in priority order (first wins, in event of a tie, outside of override configuration). The processor may need to be updated to allow the override flag to apply to each individual detector.

The rules for attributes would follow entity merging rules, as defined for the SDK resource proivder.

Note: While this proposals shows a new processor replacing the `resourcedetection` processor, the details of whether to modify-in-place the existing `resourcedetection` processor or create a new one would be determined as a follow up to this design. Ideally, we don't want users to need new configuration for resource in the otel collector.

## Datamodel Changes

Given our desired design and algorithms for detecting, merging and manipulating Entities, we need the ability to denote how entity and resource relate. These changes must not break existing usage of Resource, therefore:

> - The Entity model must be *layered on top of* the Resource model.  A system does not need to interact with entities for correct behavior.

Reviewer 14: Does the opt-in nature here conflict with the goals of unblocking the pieces you mention at the start?  I'm wondering specifically about some components in a pipeline using the entities while others are only using resource.  In such a case, I would expect that two entities could appear in the pipeline because now we have two areas of reference for the same concept.

If entiies becomes a new scope, should this eventually seek to replace resource?

Author: I think entities will eventually replace existing resource usage.   The reason we're layering is to not break the entire ecosystem and effectively create a V2 of the protocol.  Instead, we see this as an evolution.  If you want to engage with Entities, you would NEED to ensure all components are entity aware.  As soon as one component is not, you're back to today's behavior.  So users who want to use entities MUST ensure all components leverage entities.

However - there's a higher level question.  Do I need to understand entities to interact with Metrics/Logs/Spans/etc. from OpenTelemetry?  The answer to that is no.  You can blindly leverage `Resource.attributes` as important slices of information for the data, and you will always be able to continue doing that.  That's addressing consumers / storage solutions.

### How to deal with Resource/Entities whose lifecycle does not match the SDK?

This proposal motivates a Resource Provider in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.

### How to deal with Prometheus Compatibility for non-SDK telemetry?

Today, [Prometheus compatibility](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md) relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.

Here's a list of requirements for the solution:

- Existing prometheus/OpenTelemetry users should be able to migrate from where they are today.

> - Prometheus should have a set of identifying attributes for their up and coming `info()` function and info metric features.

Reviewer 13: Please remember that the upcoming `info` function MVP doesn't change the info metrics themselves, i.e. it still needs to rely on convention for which are the `target_info` identifying labels (i.e. `instance` and `job`). Furthermore, the Prometheus OTLP endpoint still needs to be able to generate `instance` and `job` labels for `target_info`.

Author: I'll update this to be explicit about that.

Author: added!

> - Prometheus should have a set of identifying attributes for their up and coming `info()` function and info metric features.

Reviewer 15: I'd like to propose something different here. 

Let's first remember the motivation behind `target_info`. Resource attributes, before this OTEP, can be a gigantic array of information and not every attribute makes sense to become a metric label in the Prometheus world. Instead of adding all attributes to all metrics, which increases network bandwidth unnecessarily (almost all resources wouldn't be used at all), everything is encoded to `target_info` instead[[1](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#:~:text=A%20Prometheus%20Info%20metric%20MUST%20be%20converted%20to%20an%20OTLP%20Non%2DMonotonic%20Sum%20unless%20it%20is%20the%20target%20info%20metric%2C%20which%20is%20used%20to%20populate%20resource%20attributes)]. 

Once `target_info` became a thing, PromQL Joins were the way to decode resource attributes from `target_info` and use them in queries with other metrics. Several learnings from this pattern eventually led to the creation of the `info()` function! The `info()` function is a massive UX improvement when compared to the usage of joins, but let's not forget the original intention of all of this: we want an efficient way to include just useful OTel attributes as metric labels.

I'm not sure if I'm missing something, but optimizing this proposal to make sure info() and identifying attributes work well together seems a bit like the [XY problem](https://www.google.com/url?sa=i&url=https%3A%2F%2Ftwitter.com%2Fsketchplanator%2Fstatus%2F1743988378863800608&psig=AOvVaw2hueW4eWExr97F9pYhF9_D&ust=1726763312621000&source=images&cd=vfe&opi=89978449&ved=0CBQQjRxqFwoTCOjy4-f0zIgDFQAAAAAdAAAAABAE).

--- 

This OTEP proposes a separation of resource attributes: identifying and descriptive attributes. Would it make sense to just add all identifying attributes as metric labels? 

Identifying attributes are immutable and, hopefully, a much shorter set when compared to the whole resource attribute set. This solves the problem of Prometheus receiving a gigantic set of labels that would never be used.

If descriptive attributes are still desired by users for query filters, would it make that the `prometheusexporter` and `prometheusremotewriteexporter` (otelcontribcol components) have configuration options that allow promoting particular descriptive attributes as labels? Additionally, we can propose that Prometheus /otlp endpoint can also promote particular descriptive attributes as labels.

Reviewer 8: > If descriptive attributes are still desired by users for query filters, [...]
I think this statement is the core misconception, here. The identifying attributes are the shortest path to identify the entity technically, but it is not necessarily enough for a human to understand what that entity really is. It might be some extra technical ID on top of identifying service attributes that might not even make sense to the end user receiving the data. I can share the example that brought me to this discussion if it helps.

Reviewer 13: Agree with @feldentm-SAP. I think that bringing OTel resource attribute to Prometheus label promotion into this proposal is a distraction. As I see it, whether/how to promote resource attributes should be layered on top of basic OTel/Prometheus interop, which is what should be covered in the Entity proposal.

Reviewer 15: Hmmm, okay. My understanding was that it would be rare for people to want descriptive attributes in their queries.

Would it make sense to list what you predict to be common identifying and descriptive attributes in the proposal?

Reviewer 13: > Would it make sense to list what you predict to be common identifying and descriptive attributes in the proposal?

Won't this be entirely domain dependent?

Reviewer 15: > Won't this be entirely domain dependent?

Yes, I think you're correct. Maybe it's too much to add to the proposal. I'm just hoping for a prediction so we can compare it to our [current recommendations](https://prometheus.io/docs/guides/opentelemetry/#promoting-resource-attributes).

Reviewer 13: Again, the concepts of identifying OTel resource attributes and which to promote to Prometheus metric labels should be kept orthogonal to each other.

Reviewer 8: Just for clarification: I expect that identifying attributes will overlap, i.e. belong to multiple entities. Because we have requirements to express sub resources. The implications of this vary depending on Prometheus' role in the observability stack. If it is used as a routing component, it should pretty much behave like an OTel collector and all available attributes must be passed through it. If it is used as data sink, a consumer might want to filter some irrelevant labels before. However, I'm not aware of this happening a lot and would, hence, not let it guide design decisions, right?

Author: I think this is an interesting proposal, but I think it's important to understand why resource exists in otel in the first place.

I wrote a bit about how it's used [here](https://docs.google.com/document/d/1Xd1JP7eNhRpdz1RIBLeA1_4UYPRJaouloAYqldCeNSc/edit).

A shortened version:

- It must identify an entity that is producing telemetry.
- It should allow users to determine where that entity resides within their infrastructure.

Prometheus metric labels *can* do this as well. But a few things I don't have in that paper that's important for OpenTelemetry:

- Resource provides "guaranteed" mechanism for correlating o11y signals by "source".
- OpenTelemetry pipelines (OTLP) leverage resource as a default bundling / aggregation boundary for signals.

When we speak of how things encode in prometheus, there's a few questions I have:

1. How will users of prometheus interact with other signals?  E.g. how can I go from a an alert on a timeseries to finding relevant logs/traces from the time-period of the anamolous behavior?  Does preserving the semantic of "resource" make this easier?
2. When federating Prometheus or interacting with OTLP, should we be able to "round trip" this semantic information both in & out?  I.e. should we be able to export OTLP from prometheus that looks (as close as possible) to the OTLP that created the timeseries in Prometheus to begin with?

In my mind - I would love if users of Prometheus and other backends for OTLP signals can consistently understand "resource.{label}" means the same thing between the signals, and can design their system around this.

In terms of our discussion, let's outline things I'm implicitly hearing agreement to, or that I think are inputs into the design space:

- `Info()` seems needed to deal with descriptive attributes, no matter what.  The proposals I've seen for `Info`, I think, solve this issue well.
- Having an arbitrary number of labels to `join` Info metrics is problematic from performance and implementation persective (@aknuds1 let me know if I'm reading too much into this).
- OpenTelemetry (today) allows an aribitrary number of identifying attributes.  Unfortunately, users are placing descriptive attributes in that location today, because of the usefulness of "resource" as a concept for x-telemetry correlation and labeling.
- OpenTelemetry *only* guarantees `job/instance` labels for SDK where we can ensure a `service.name` and `service.instance.id` are required (although that's not fully consistent across our SDKs yet, as it's "new"-ish requirement).
- Entities proposal gives us the ability to split identifying + descriptive attributes in Resource.  Additionally, we will have some notion of types of resource.  

Given this requirements, let's try to answer the two questions I listed above.  Looking forward to the discussion later today.

Reviewer 13: @jsuereth 

> How will users of prometheus interact with other signals? E.g. how can I go from a an alert on a timeseries to finding relevant logs/traces from the time-period of the anamolous behavior? Does preserving the semantic of "resource" make this easier?

I'm not the biggest expert on this, but in my personal experience exemplars are useful for finding traces corresponding to metrics. In the OTel sphere, I think that knowing the identifying resource attributes (ideally via Prometheus metadata?) would be really helpful in finding other signals from the same producer. Agree?

> When federating Prometheus or interacting with OTLP, should we be able to "round trip" this semantic information both in & out? I.e. should we be able to export OTLP from prometheus that looks (as close as possible) to the OTLP that created the timeseries in Prometheus to begin with?

I haven't given much thought to round tripping, but ideally it should be possible to re-create the OTLP losslessly. I figure that storing OTel resource attributes as metadata ought to facilitate this scenario. Make sense?

> Info() seems needed to deal with descriptive attributes, no matter what. The proposals I've seen for Info, I think, solve this issue well.

Yeah, the `info` [MVP](https://github.com/prometheus/prometheus/pull/14495) which will hopefully be approved, basically acts like a smart join query against `target_info`. So it depends on `instance` and `job` being identifying labels, plus all descriptive attributes being labels of `target_info`.

> Having an arbitrary number of labels to join Info metrics is problematic from performance and implementation persective (@aknuds1 let me know if I'm reading too much into this).

The current version of `info`, which doesn't require any new Prometheus metadata capabilities, depends on `instance` and `job` being _the_ identifying `target_info` labels. If Prometheus were to have persisted metadata, `info` could handle arbitrary identifying labels, but too many would still cause metrics to have too many labels (could hit limits) etc. It's better to keep it simple :)

Author: > I'm not the biggest expert on this, but in my personal experience exemplars are useful for finding traces corresponding to metrics.

OpenTelemetry provides (at least) two types of correlation:

- "Causal" correlation.  You *know* that "A" leads to "B".  We do this through (trace) id propagation and Exemplars are a means to track participants in the "cause" of a metric being the way it is.
- "Source" correlation.  This is what "Resource" does.  What system/component created "signal A"?  Is that the same component that's also creating problematic signal "B"?

They're different goals and both are viable/useful.  We can't only rely on a (trace) IDs for causal relationships, just like we can't only rely on causal relationships for o11y use cases.

Author: > In the OTel sphere, I think that knowing the identifying resource attributes (ideally via Prometheus metadata?) would be really helpful in finding other signals from the same producer. Agree?

Yes.

> I haven't given much thought to round tripping, but ideally it should be possible to re-create the OTLP losslessly. I figure that storing OTel resource attributes as metadata ought to facilitate this scenario. Make sense?

Yes, I think metadata should be in the design considerations here.

My main concern is the details on how this would work, and I'm relying on @aknuds1's opinion here for viability.

Author: Based on the in person proposal, I've updated this section. TL:DR;

- We continue to leverage `target_info` and `Resource.attributes` as we do today.  
- We will work together to design entities and Prometheus features to better the experience of O11y of OpenTelemetry + Prometheus.  We can progress with this proposal and hash out the prometheus<->otel design as we go.

> - (desired) OpenTelemetry should be able to create unique `job`/`instance` labels between all metrics sent to prometheus for any "info" metric join.

Reviewer 13: Should we be specific about the `target_info` metric, to avoid any ambiguity?
```suggestion
- (desired) OpenTelemetry should be able to create unique `job`/`instance` labels between all metrics sent to prometheus for any "target_info" metric join.
```

A quick proposal of what this might look like:

- `target_info` metric generation is updated to exclude any keys which are contained in `descriptive_attributes_keys` of an entity.

> - For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD), which has all identifying and descriptive keys. This should play nicely with the planned improvements to [info-typed metrics](https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals).

Reviewer 13: > For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD)

Isn't this really the same as the option in the paragraph below ("that would break compatibility")? I wonder if this paragraph instead should mention synthesizing a `target_info` `instance` label from all the identifying resource attributes.

Another option (that would break compatibility):

- `job`/`instance` are synthesized using a consistent hashing algorithm and identifying attributes of entity on resource.

> - Each entity type is written as an info metric called `<entity_type>_entity-info`.

Reviewer 13: ```suggestion
- Each entity type is written as an info metric called `<entity_type>_entity_info`.
```

Author: Fixed

| descriptive_attributes_keys | repeated string | Descriptive (non-identifying) attribute keys of the entity.<br/>MAY change over the lifetime of the entity. MAY be empty. These attribute keys are not part of entity's identity.<br/><br/>These keys MUST exist in Resource.attributes.<br/><br/>Follows any value definition in the OpenTelemetry spec - it can be a scalar value, byte array, an array or map of values. Arbitrary deep nesting of values for arrays and maps is allowed.<br/><br/>SHOULD follow OpenTelemetry semantic conventions for attributes.| now a reference |

### Resource Identity

OpenTelemetry resource identity will be modified as follows:

- When `entities` is empty on resource, then its identity is the collection
  of all `attributes` (both key and values).
- When `entities` is non-empty on resource, then its identity is the collection

>   of all `attributes` where the key is found in `entities.identify_attribute_keys`.

Author: I like this, I think it gives us more scenarios where entity can provide value.

>   of all `attributes` where the key is found in `entities.identify_attribute_keys`.

Reviewer 8: To me, it would be obscure to reverse the logic because a) I would expect the important attributes being marked (over marking the unimportant ones and stating that important attributes are unmarked attributes) b) I would not expect to explicitly state which attributes are unimportant especially because I would expect the number of identifying attributes to be significantly smaller.

>   of all `attributes` where the key is found in `entities.identify_attribute_keys`.

Reviewer 13: I think I agree with @feldentm-SAP. Would it be a showstopper to require defining the identifying attributes when adding the first entity to a resource? It seems like a bit more of a footgun if identifying attributes are implicitly defined, as the inverse of the non-identifying attribute set.

### How to deal with Resource/Entities whose lifecycle does not match the SDK?

This proposal motivates a Resource Provider in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.

### How to deal with Prometheus Compatibility for non-SDK telemetry?

Today, [Prometheus compatibility](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md) relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.

Here's a list of requirements for the solution:

- Existing prometheus/OpenTelemetry users should be able to migrate from where they are today.

> - Any solution MUST work with the [info-typed metrics](https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals) being added in prometheus.

Reviewer 15: @aknuds1 since the proposal is being referenced in other places, like this OTEP, I wonder if we should update the proposal with what is being implemented [here](https://github.com/prometheus/prometheus/pull/14495). 🤔

  - Resource descriptive attributes should leverage `info()` or metadata.
  - Resource identifying attributes need more thought/design from OpenTelemetry semconv + entities WG.
  - Note: Current `info()` design will only work with `target_info` metric, and `job/instance` labels for joins.  This labels MUST be generated by the OTLP endpoint in prometheus.
- (desired) Users should be able to correlate metric timeseries to other signals via Resource attributes showing up as labels in prometheus.

> - (desired) Conversion from `OTLP -> prometheus` can be reversed such that `OTLP -> Prometheus -> OTLP` is non-lossy.

Reviewer 15: > - (desired) Conversion from `OTLP -> prometheus` can be reversed such that `OTLP -> Prometheus -> OTLP` is non-lossy.

I'm bringing that up to the Prometheus team. We've recently decided that [Prometheus won't support exporting OTLP directly](https://github.com/prometheus/prometheus/issues/12633#issuecomment-2359054297), but we're working on [Prometheus Remote Write Receiver in OTel collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35295) and that's our plan to transform Prometheus metrics into OTLP format, would that be ok?

Author: Yes - to me it's the same domain modelling issue

- **Embed fully Entity in Resource.** This was rejected because it makes it easy/trivial for Resource attributes and Entities to diverge.  This would prevent the backwards/forwards compatibility goals and also require all participating OTLP users to leverage entities. Entity should be an opt-in / additional feature that may or may not be engaged with, depending on user need.
- **Re-using resource detection as-is** This was rejected as not having a viable compatibility path forward.  Creating a new set of components that can preserve existing behavior while allowing users to adopt the new functionality means that users have better control of when they see / change system behavior, and adoption is more obvious across the ecosystem.

## Future Posibilities

> This proposal opens the door for addressing issues where an Entity's lifetime does not match an SDK's lifetime, in addition to providing a data model where mutable (descriptive) attributes can be changed over the lifetime of a resource without affecting its identity.  We expect a follow-on OTEP which directly handles this issue.

Reviewer 16: Two things I would see valuable to be addressed going forward as well:

* (Optional) attributes per resource/entity to identify where it was set. This is especially useful in the cases described above where multiple places in the flow try to set a resource/entity and one wins over the other. Imagine that I configure a collector with a processor that should set some k8s.pod.id, but for whatever reason the id is not what I expect, because I am not aware that the id is set by the k8s operator via an environment variable. Debugging that would become a mess, so having an (opt in?) attribute that annotates where the data is coming from would be helpful.
* Maybe as part of the semantic conventions a definition on how certain attributes should be collected. E.g. we had a discussion in the past around [`host.id`](https://opentelemetry.io/docs/specs/semconv/attributes-registry/host/), what the value should be and where it should come from.

Author: (2) is already called out and on the task queue for Semantic Conventions + Entities WG.

(1) Is an interesting idea. I think there will be a whole "debugability" thing to figure out, but honestly the scenario you mention is already an issue today.  It's an interesting idea to call out as a potential solution or "diagnostic" mode we could support.
